File: src/pysimdata_cplus/generators/bilateral_gaussian.cpp
#include "pysimdata_cplus/generators/bilateral_gaussian.h"

#include <cmath>

namespace pysimdata_cplus {

BilateralGaussian::BilateralGaussian(int rows, int cols, double sigma,
                                     double amplitude, int center_y, int center_x)
    : rows_(rows), cols_(cols), sigma_(sigma), amplitude_(amplitude),
      center_y_(center_y), center_x_(center_x) {
    params_ = {
        {"shape", json::array({rows_, cols_})},
        {"sigma", sigma_},
        {"amplitude", amplitude_},
        {"center", json::array({center_y_, center_x_})}
    };
}

BilateralGaussian::BilateralGaussian(const json& params) {
    auto shape = GetPairParam<int>(params, "shape", {256, 256});
    rows_ = shape.first;
    cols_ = shape.second;
    sigma_ = GetParam<double>(params, "sigma", 10.0);
    amplitude_ = GetParam<double>(params, "amplitude", 255.0);
    auto center = GetPairParam<int>(params, "center", {-1, -1});
    center_y_ = center.first;
    center_x_ = center.second;
    params_ = params;
}

json BilateralGaussian::to_config() const {
    return {
        {"type", "BilateralGaussian"},
        {"format", "csv"},
        {"params", params_}
    };
}

Eigen::MatrixXd BilateralGaussian::do_generate() {
    int cy = (center_y_ < 0) ? rows_ / 2 : center_y_;
    int cx = (center_x_ < 0) ? cols_ / 2 : center_x_;

    Eigen::MatrixXd result(rows_, cols_);
    for (int i = 0; i < rows_; ++i) {
        for (int j = 0; j < cols_; ++j) {
            double dist = std::abs(i - cy) + std::abs(j - cx);
            result(i, j) = amplitude_ * std::exp(-dist / sigma_);
        }
    }
    return result;
}

}  // namespace pysimdata_cplus

File: src/pysimdata/function/bilateral_gaussian.py
"""双边高斯分布生成器"""
from typing import Optional, Tuple

import numpy as np

from ..base import BaseGenerator


class BilateralGaussian(BaseGenerator):
    """双边高斯分布生成器"""

    CONFIG_KEYS = {
        "shape": ("shape", tuple),
        "sigma": ("sigma", float),
        "amplitude": ("amplitude", float),
        "center": ("center", tuple),
    }

    def __init__(
        self,
        shape: Tuple[int, int] = (256, 256),
        sigma: float = 10.0,
        amplitude: float = 255.0,
        center: Optional[Tuple[int, int]] = None,
        data_source: Optional[dict] = None,
    ):
        def _func(shape, sigma, amplitude, center):
            if center is None:
                center = (shape[0] // 2, shape[1] // 2)
            y, x = np.ogrid[:shape[0], :shape[1]]
            y = y - center[0]
            x = x - center[1]
            dist = np.abs(y) + np.abs(x)
            return amplitude * np.exp(-dist / sigma)

        super().__init__(
            func=_func,
            data_source=data_source,
            shape=shape,
            sigma=sigma,
            amplitude=amplitude,
            center=center,
        )